hil/ext/switches/dellnos9.py

```diff
@@ -0,0 +1,377 @@
+
+Uses the XML REST API for communicating with the switch.
+"""
+
+import logging
+from lxml import etree
+import re
+import requests
+import schema
+
+from hil.model import db, Switch, SwitchSession
+from hil.errors import BadArgumentError
+from hil.model import BigIntegerType
+from hil.network_allocator import get_network_allocator
+
+logger = logging.getLogger(__name__)
+
+CONFIG = 'config-commands'
+SHOW = 'show-command'
+
+
+class DellNOS9(Switch, SwitchSession):
+    """Dell S3048-ON running Dell NOS9"""
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/dellnos9'
+
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(BigIntegerType,
+                   db.ForeignKey('switch.id'), primary_key=True)
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+    interface_type = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        schema.Schema({
+            'hostname': basestring,
+            'username': basestring,
+            'password': basestring,
+            'interface_type': basestring,
+        }).validate(kwargs)
+
+    def session(self):
+        return self
+
+    @staticmethod
+    def validate_port_name(port):
+        """Valid port names for this switch are of the form 1/0/1 or 1/2"""
+
+        val = re.compile(r'^\d+/\d+(/\d+)?$')
+        if not val.match(port):
+            raise BadArgumentError("Invalid port name. Valid port names for "
+                                   "this switch are of the from 1/0/1 or 1/2")
+        return
+
+    def disconnect(self):
+        pass
+
+    def modify_port(self, port, channel, new_network):
+        (port,) = filter(lambda p: p.label == port, self.ports)
+        interface = port.label
+
+        if channel == 'vlan/native':
+            if new_network is None:
+                self._remove_native_vlan(interface)
+            else:
+                self._set_native_vlan(interface, new_network)
+        else:
+            legal = get_network_allocator(). \
+                is_legal_channel_for(channel, new_network)
+            assert legal, "HIL passed an invalid channel to the switch!"
+            vlan_id = channel.replace('vlan/', '')
+
+            if new_network is None:
+                self._remove_vlan_from_trunk(interface, vlan_id)
+            else:
+                assert new_network == vlan_id
+                self._add_vlan_to_trunk(interface, vlan_id)
+
+    def revert_port(self, port):
+        self._remove_all_vlans_from_trunk(port)
+        if self._get_native_vlan(port) is not None:
+            self._remove_native_vlan(port)
+
+    def get_port_networks(self, ports):
+        response = {}
+        for port in ports:
+            native = self._get_native_vlan(port.label)
+            # return an empty list if native is None
```

Comment:
Better to just change the expression to make this more obvious; options:

* `native = [native] if native is not None else []`
* `native = [] if native is None else [native]`
* An if-statement, rather than expression:

```python
if native is None:
    native = []
else:
    native = [native]
```

Reply:
Better yet, you could just do:

```python
response[port] = self._get_vlans(port.label)
if native is not None:
    response[port].append(native)
```